src/parse.rs

```diff
@@ -877,16 +991,22 @@ impl<'a> Bytes<'a> {
         }
     }
 
-    fn parse_escape(&mut self) -> Result<char> {
+    fn parse_escape(&mut self, encoding: EscapeEncoding) -> Result<EscapeCharacter> {
         let c = match self.eat_byte()? {
-            b'\'' => '\'',
-            b'"' => '"',
-            b'\\' => '\\',
-            b'n' => '\n',
-            b'r' => '\r',
-            b't' => '\t',
-            b'0' => '\0',
-            b'x' => self.decode_ascii_escape()? as char,
+            b'\'' => EscapeCharacter::Ascii(b'\''),
+            b'"' => EscapeCharacter::Ascii(b'"'),
+            b'\\' => EscapeCharacter::Ascii(b'\\'),
+            b'n' => EscapeCharacter::Ascii(b'\n'),
+            b'r' => EscapeCharacter::Ascii(b'\r'),
+            b't' => EscapeCharacter::Ascii(b'\t'),
+            b'0' => EscapeCharacter::Ascii(b'\0'),
+            b'x' => {
+                let b = self.decode_ascii_escape()?;
+                match encoding {
+                    EscapeEncoding::Binary => EscapeCharacter::Ascii(b),
+                    EscapeEncoding::Utf8 => EscapeCharacter::Utf8(b as char),
```

Comment:
This is wrong: `\x..` may not be a valid UTF8 character by itself - we need to parse further bytes to find out if a valid character is found or not